packages/subscription/src/semaphore.ts

```diff
@@ -0,0 +1,12 @@
+export class Semaphore<T> {
+  waiters: Array<(value: T) => void> = [];
+
+  signal = (value: T) => {
+    let next = this.waiters.pop();
+    if (next) {
+      next(value);
+    }
+  }
+
+  wait = () => new Promise<T>(resolve => this.waiters.push(resolve));
+}
```

Comment:
This is so cool, I've though about whether a semaphore/condvar is a thing that could be a low level primitive for us, and this is such a neat and simple implementation. Love it!